cpukit/score/src/threadprioritynode.c

```diff
@@ -0,0 +1,184 @@
+ */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <rtems/system.h>
+#include <rtems/config.h>
+#include <rtems/score/chain.h>
+#include <rtems/score/coremutex.h>
+#include <rtems/score/coremuteximpl.h>
+#include <rtems/score/thread.h>
+#include <rtems/score/threadimpl.h>
+
+/**
+ * This routine adds a thread's Priority_node to the mutex holder's
+ * Inherited_priorities list, and sets the waiting_to_hold field to the
+ * mutex.
+ *
+ * Returns true if the enqueued node is at the head of the holder's
+ * Inherited priorities, and therefore the holder's priority may change.
+ */
+bool _Thread_Enqueue_priority_node(
+  Thread_Control     *the_thread,
+  CORE_mutex_Control *the_mutex
+)
+{
+  Chain_Node *iter;
+  Thread_Priority_node *thread_pn, *holder_pn;
+  Priority_Control task_priority, iter_priority;
+  
+  thread_pn = &the_thread->Priority_node;
+  holder_pn = &the_mutex->holder->Priority_node;
+
+  thread_pn->waiting_to_hold = the_mutex;
+  iter = _Chain_First( &holder_pn->Inherited_priorities );
+  task_priority = thread_pn->current_priority;
+
+  /* Find the first node in holder's Inherited_priorities list with
+   * lower priority than the task, or find the end of the list. */
+  while ( !_Chain_Is_tail( &holder_pn->Inherited_priorities, iter) ) {
+    iter_priority = ((Thread_Priority_node *)iter)->current_priority;
+    if ( iter_priority > task_priority ) {
+      break;
+    }
+    iter = _Chain_Next( iter );
+  }
+  _Chain_Insert_unprotected( _Chain_Previous( iter ), &thread_pn->Node );
+
+  /* If the_thread is at the start of the holder's Inherited_priorites list,
+   * then may need to update the holder's current priority. */
+  if ( _Chain_Is_first( &thread_pn->Node ) ) {
+    return true;
+  }
+  return false;
+}
+
+/**
+ * This routine sets the thread's priority_node->current_priority to the
+ * minimum of the real_priority and the first node of Inherited_priorities,
+ * if any.
+ */
+void _Thread_Evaluate_priority(
+  Thread_Control *the_thread
+)
+{
+  Chain_Node *first;
+  Thread_Priority_node *first_pn, *priority_node;
+  Priority_Control current_priority, min_priority;
+
+  /* Start with the_thread, and iterate through the holder threads that
+   * the_thread is waiting on. If any thread's priority does not change,
+   * then stop iterating. */
+  while ( the_thread != NULL ) {
+    priority_node = &the_thread->Priority_node;
+    current_priority = priority_node->current_priority;
+    min_priority = priority_node->real_priority;
+    first = _Chain_First( &priority_node->Inherited_priorities );
+    if ( !_Chain_Is_tail( &priority_node->Inherited_priorities, first ) ) {
+      first_pn = (Thread_Priority_node *)first;
+      if ( min_priority > first_pn->current_priority ) {
+        min_priority = first_pn->current_priority;
+      }
+    }
+    if ( current_priority < min_priority ) {
+        _Thread_Restore_priority(
+                the_thread
+                );
```

Comment:
lw

Reply:
also, i'm not sure the line wrapping is needed here. the line shouldn't be too long I believe.